### PR TITLE
Issue1314 bestest weather

### PR DESCRIPTION
**Pull request by EttoreZ**
This issue deals with ASHRAE BESTEST for weather data. This is an update in which the BESTEST was implemented in the library together with an automated way to run the tests. The folders are reported below: 
-weather data folder:_IBPSA/Resources/Data/BoundaryConditions/WeatherData/Validation
- script folder: IBPSA/Resourcessrc/BoundaryConditions/Validation/BESTEST

### CODE REVIEW

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 1]
```diff
@@ -0,0 +1,1718 @@
...
+          __Dymola_NumberOfIntervals=35043,
+          Tolerance=1e-06,
+          __Dymola_Algorithm="Dassl"),
+    __Dymola_Commands(file="modelica://IBPSA/Resources/Scripts/Dymola/BoundaryConditions/SolarGeometry/Examples/IncidenceAngle.mos"
+            "Simulate and plot"));
+    end WD600;
+
+  end Validation;
+
+  model ComRadIsoPerDir
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:1641]: Please use a full name (camel case is used for instances, not for class names).

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 2]
```diff
@@ -0,0 +1,1718 @@
...
+<li>Humidity ratio (kg moisture/kg dry air)</li>
+<li>Wet bulb temperature (&deg;C)</li>
+<li>Windspeed (m/s)</li>
+<li>Wind direction (degrees from north)</li>
+<li>Station pressure (mbar)</li>
+<li>Total cloud cover (tenths of sky)</li>
+<li>Opaque cloud cover (tenths of sky)</li>
+<li>Sky temperature (&deg;C)</li>
+</ul>
+<p><br>&bull; Sum of total, beam, and diffuse solar radiation incident on each surface (Wh/m2)&nbsp;</p>
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:84]: Use `<ul><li>` rather than `<br>&bull;`

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 3]
```diff
@@ -0,0 +1,1718 @@
...
+    model WD100 "Test model for BESTEST weather data: base case"
+      extends Modelica.Icons.Example;
+      parameter Modelica.SIunits.Angle lat=0.6952170009469 "Latitude angle";
+      parameter Real rho=0 "Ground reflectance";
+
+      WeatherData.ReaderTMY3                          weaDat(
+        pAtmSou=IBPSA.BoundaryConditions.Types.DataSource.File,
+        ceiHeiSou=IBPSA.BoundaryConditions.Types.DataSource.Parameter,
+        ceiHei=1650,
+        totSkyCovSou=IBPSA.BoundaryConditions.Types.DataSource.File,
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:150]: `File` is the default, so no need to declare it here.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 4]
```diff
@@ -0,0 +1,1718 @@
...
+        HSou=IBPSA.BoundaryConditions.Types.RadiationDataSource.File,
+        filNam=Modelica.Utilities.Files.loadResource(
+            "modelica://IBPSA/Resources/Data/BoundaryConditions/WeatherData/Validation/725650.mos"),
+        calTSky=IBPSA.BoundaryConditions.Types.SkyTemperatureCalculation.HorizontalRadiation)
+        annotation (Placement(transformation(extent={{-10,-10},{10,10}},
+            rotation=90,
+            origin={0,-88})));
+
+      WeatherData.Bus weaBus annotation (Placement(transformation(extent={{-16,-82},
+                {14,-54}}), iconTransformation(extent={{-220,70},{-200,90}})));
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:168]: Add comment.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 5]
```diff
@@ -0,0 +1,1718 @@
...
+  end UsersGuide;
+
+  package Validation
+    "Boundary conditions validation according to BESTEST specifications"
+    model WD100 "Test model for BESTEST weather data: base case"
+      extends Modelica.Icons.Example;
+      parameter Modelica.SIunits.Angle lat=0.6952170009469 "Latitude angle";
+      parameter Real rho=0 "Ground reflectance";
+
+      WeatherData.ReaderTMY3                          weaDat(
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:146]: Fix white space.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 6]
```diff
@@ -0,0 +1,1718 @@
...
+
+      WeatherData.Bus weaBus annotation (Placement(transformation(extent={{-16,-82},
+                {14,-54}}), iconTransformation(extent={{-220,70},{-200,90}})));
+      ComRadIsoPerDir Azi000Til00(
+        til=IBPSA.Types.Tilt.Ceiling,
+        lat=lat,
+        azi=IBPSA.Types.Azimuth.S,
+        rho=rho) "Azimuth = Horizontal, Tilt = 0 °"
+        annotation (Placement(transformation(extent={{68,68},{88,88}})));
+      ComRadIsoPerDir Azi000Til90(
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:175]: Use lower case for instance names, e.g., `azi00....`

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 7]
```diff
@@ -0,0 +1,1718 @@
...
+        HInfHorSou=IBPSA.BoundaryConditions.Types.DataSource.File,
+        HSou=IBPSA.BoundaryConditions.Types.RadiationDataSource.File,
+        filNam=Modelica.Utilities.Files.loadResource(
+            "modelica://IBPSA/Resources/Data/BoundaryConditions/WeatherData/Validation/725650.mos"),
+        calTSky=IBPSA.BoundaryConditions.Types.SkyTemperatureCalculation.TemperaturesAndSkyCover)
+        annotation (Placement(transformation(extent={{-10,-10},{10,10}},
+            rotation=90,
+            origin={50,-90})));
+
+      WeatherData.Bus weaBusDew annotation (Placement(transformation(extent={{36,-82},
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:254]: I don't see why you need `weaBusDew` and the data reader that outputs the dew point.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 8]
```diff
@@ -0,0 +1,1718 @@
...
+          StopTime=31539600,
+          __Dymola_NumberOfIntervals=35043,
+          Tolerance=1e-06,
+          __Dymola_Algorithm="Dassl"),
+    __Dymola_Commands(file="modelica://IBPSA/Resources/Scripts/Dymola/BoundaryConditions/SolarGeometry/Examples/IncidenceAngle.mos"
+            "Simulate and plot"));
+    end WD100;
+  extends Modelica.Icons.ExamplesPackage;
+
+    model WD200
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:390]: I have not looked in detail into the test specification, but `WD200` seems very similar to `WD100`. If this is the case, simply extend `WD100` and assign the parameters that are different and add additional models if any are needed only for `WD200`. Copying code makes maintenance hard and makes it very difficult to see why some tests may work but not others.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 9]
```diff
@@ -0,0 +1,1718 @@
...
+            -82},{-8,-30},{48,-30}}, color={0,0,127}));
+    connect(HDiffIso.H, AddHdirHdiffIso.u1) annotation (Line(points={{-11,72},{2,72},
+            {2,24},{46,24}}, color={0,0,127}));
+    connect(HDir.H, AddHdirHdiffIso.u2)
+      annotation (Line(points={{-11,0},{2,0},{2,12},{46,12}}, color={0,0,127}));
+    connect(AddHdirHdiffIso.y, H) annotation (Line(points={{69,18},{86,18},{86,0},
+            {110,0}}, color={0,0,127}));
+    connect(AddHdirHdiffPer.y, HPer) annotation (Line(points={{71,-24},{86,-24},{86,
+            -26},{110,-26}}, color={0,0,127}));
+    annotation (Icon(coordinateSystem(preserveAspectRatio=false)), Diagram(
```
Comment (mwetter) [inline IBPSA/BoundaryConditions/BESTEST.mo:1710]: This block must have an info and revision section. (We require this from all classes.)

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 1 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:1641]: Added full name "IsotropicAndPerezDiffuseRadiation" instead of camelcase

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 10]
```diff
@@ -0,0 +1,1718 @@
+﻿within IBPSA.BoundaryConditions;
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:1]: I see, now it should be fine

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 11]
```diff
@@ -0,0 +1,1718 @@
...
+<td><p>30&deg; from horizontal</p></td>
+</tr>
+<tr>
+<td><p>West</p></td>
+<td><p>30&deg; from horizontal</p></td>
+</tr>
+</table>
+<p><br><h4>Additional parameters and correlations:</h4></p>
+<p>Ground reflectance &rho; can be set to 0 or 0.2</p>
+<p>Black body temperature calculated using Horizontal radiation or dew point temperature and sky cover</p><p>Perez or Isoentropic sky model used</p>
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:59]: I updated the documentation and corrected the typos. I  forgot to ask though, what do you mean by shorter lines?

File: IBPSA/Resources/src/BoundaryConditions/Validation/BESTEST/WeatherBESTEST.py [hunk 12]
```diff
@@ -0,0 +1,815 @@
...
+CLEAN_MAT = True
+# Erse anything but the Json file results in the ResultJson folder and .matfiles
+DelEvr = False
+#Get IBPSA library from gitHub
+FROM_GIT_HUB = False
+# Modelica IBPSA Library working branch
+#BRANCH = 'issue1314_BESTEST_weather'
+BRANCH ='master'
+#Path to local library copy
+LIBPATH = "C:\\Users\\ettor\\Documents\\Repositories\\MODELICA\\TRY\\modelica-ibpsa\\IBPSA"
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/Resources/src/BoundaryConditions/Validation/BESTEST/WeatherBESTEST.py:32]: I thought since it could be helpful to have a local pathing, instead of having it hardwired it refers to the relative path inside the repository folder using "pathlib"

File: IBPSA/Resources/src/BoundaryConditions/Validation/BESTEST/WeatherBESTEST.py [hunk 13]
```diff
@@ -0,0 +1,815 @@
...
+ 'toDryAir.XiDry', 'weaBus.pAtm', 'weaBus.TDryBul', 'weaBus.relHum', 'weaBus.TBlaSky', \
+ 'weaBus.TDewPoi', 'weaBus.TWetBul', 'weaBus.nOpa', 'weaBus.nTot', 'weaBus.winDir', \
+ 'weaBus.winSpe', 'weaBusDew.TBlaSky','Azi270Til30.HPer','Azi045Til90.HPer','Azi090Til30.HDiffIso.H']
+
+
+
+#Software specifications 
+modeler_organization = 'LBNL'
+modeler_organization_for_tables_and_charts ='LBNL'
+program_name_and_version = 'IBPSA v4.0.0dev commit: 44c8cb5446f16e6f241494fb47f7d61418248f69, Buildings v7.0.0dev 56a796c3f2b93d86e5ea8c9aa3084457b160a478'
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/Resources/src/BoundaryConditions/Validation/BESTEST/WeatherBESTEST.py:63]: I included it. However, I could not find a way to get repository information without cloning it, and since I am not using the Buildings library I removed it from the program_name_and_version tag. However, I made a few changes so that it should be easier to modify the script for the Buildings library

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 14]
```diff
@@ -0,0 +1,1718 @@
...
+First implementation.
+</li>
+</ul>
+</html>"),
+    experiment(
+          StopTime=31539600,
+          __Dymola_NumberOfIntervals=35043,
+          Tolerance=1e-06,
+          __Dymola_Algorithm="Dassl"),
+    __Dymola_Commands(file="modelica://IBPSA/Resources/Scripts/Dymola/BoundaryConditions/SolarGeometry/Examples/IncidenceAngle.mos"
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:385]: I was messing around with the Dymola commands and I think that remained there for some reason, I removed it.

File: IBPSA/BoundaryConditions/BESTEST.mo [hunk 15]
```diff
@@ -0,0 +1,1718 @@
...
+<ul>
+<li>
+March 11, 2020, by Ettore Zanetti:<br/>
+First implementation.
+</li>
+</ul>
+</html>"),
+    experiment(
+          StopTime=31539600,
+          __Dymola_NumberOfIntervals=35043,
```
[the remark this answers is not in the document]
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:382]: I was using it as a reference to check wether the extrapolation done by dassl with the adaptive time step and the fixed runge kutta that I am using in the Python scripts gave the same results. It is not needed though, since in the end all the instructions for the simulation are passed in the Python script. I removed it.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 2 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:84]: Changed the bullet points.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 3 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:150]: left the default option where it was file.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 4 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:168]: comments added the whole model should be clearer

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 5 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:146]: I fixed the white spaces , sometimes though after I already fixed them, by saving the package again the spacing will become weird again. It also happens with the documentation

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 6 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:175]: Corrected all the radiation models names

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 7 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:254]: I guess this was not clear at all, the reason for that data reader it is just to read the Sky temperature using the temperatures and sky cover. I now updated with comments and changed the nomeclature making it similar to the one in the package sky temperature, so it should be more clear.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 8 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:390]: You are absolutely right, I made WD100 the reference model and extended all the others from it.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 9 above)
Comment (EttoreZ) [reply IBPSA/BoundaryConditions/BESTEST.mo:1710]: I do not understand which block specifically you are referring to, but I added info and revision for all of them now

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 11 above)
[the remark this answers is not in the document]
Comment (mwetter) [reply IBPSA/BoundaryConditions/BESTEST.mo]: By shorter lines I mean lines that are not too long, as long lines don't fully show on github's diff which makes review harder.

File: IBPSA/BoundaryConditions/BESTEST.mo (hunk 5 above)
Comment (mwetter) [reply IBPSA/BoundaryConditions/BESTEST.mo]: In dymola I suggest to simply save the file but not the whole package to avoid that the formatting gets changed again.
